### PR TITLE
refactor(lsp): omnifunc and add auto_complete

### LINKED ISSUE #25177: LSP: completion via omnifunc deletes previous character
### Problem

Using the built in completion (omnifunc) with an LSP server deletes the previous character upon completion.

https://github.com/neovim/neovim/assets/7882157/fe4bb4a6-0a65-4479-8963-e136fadd22cc

### Steps to reproduce

1. Install language server of your choice (I tested with `clangd` and `rust_analyzer`)
2. `nvim -u NONE <test file of your choice>`
3. `:lua vim.lsp.start_client{ cmd = {"/path/to/language/serer"} }`
4. `:lua vim.lsp.buf_attach_client(vim.fn.bufnr(), 1)`
5. Trigger completion via ctrl-x + ctrl-o, notice how the character right before the cursor gets deleted.

### Expected behavior

Character before the cursor is not deleted

### Neovim version (nvim -v)

v0.10.0-dev-1113+gf5a09f1b03

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

Arch Linux

### Terminal name/version

foot

### $TERM environment variable

foot

### Installation

AUR

### PR DESCRIPTION
**Pull request by glepnir**
improve our completion experiences unbloack after #24723 merge

- move to completion module and rewrite
- add new api `vim.lsp.auto_complete` for auto completion usage
  `on_attach = function(client,bufnr) vim.lsp.auto_complete(client, bufnr, fuzzy_match)`

https://github.com/neovim/neovim/assets/41671631/94820aac-ea91-4a71-8d3f-122d5a9525ef

Fix #25177  (same  https://github.com/neovim/neovim/pull/17051, https://github.com/neovim/neovim/issues/22803)

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -669,14 +669,24 @@ local function sort_completion_items(items)
   end)
 end
 
+local function charidx_without_compchar(ltext, col)
```
Comment (glepnir) [inline runtime/lua/vim/lsp/util.lua:672]: there still need respect `client.offset_encoding` right ?

File: runtime/lua/vim/lsp.lua [hunk 2]
```diff
@@ -2346,16 +2345,17 @@ function lsp.omnifunc(findstart, base)
...
-        local matches = util.text_document_completion_list_to_complete_items(result, prefix)
+        -- local encoding = client.offset_encoding
+        -- local candidates = util.extract_completion_items(result)
+        -- local curstartbyte = adjust_start_col(pos[1], line, candidates, encoding)
+        -- if startbyte == nil then
+        --   startbyte = curstartbyte
+        -- elseif curstartbyte ~= nil and curstartbyte ~= startbyte then
+        --   startbyte = match_pos
+        -- end
+        -- local prefix = startbyte and line:sub(startbyte + 1) or line_to_cursor:sub(match_pos)
```
Comment (glepnir) [inline runtime/lua/vim/lsp.lua:2356]: prefix maybe still need , let me try a new way tomorrow

**Comment by glepnir** [comment]
could we move `omnifunc` to a single file ？Increasing function parameters sideways is a bit ugly. Avoid lsp.lua becoming bloated

**Comment by justinmk** [comment]
> could we move omnifunc to a single file ？Increasing function parameters sideways is a bit ugly.

I don't understand this.

> Avoid lsp.lua becoming bloated

Spreading things around in more files cause a different problem. What is "bloated" about a since omnifunc() function in lsp.lua?

**Comment by glepnir** [comment]
The implementation of filter text and start_col on the previous code will involve file crossing. And it doesn't make sense to put some functions in util i think.  now I add `completion.lua` for `omnifunc` and support `filterText` with fuzzymatch.

**Comment by mfussenegger** [comment]
> Spreading things around in more files cause a different problem. What is "bloated" about a since omnifunc() function in lsp.lua?
> If something is bloated, code should be deleted. Spreading it out just hides the problem.


I think a lsp/completion.lua would be okayish if we use it to eliminate the existing spread. Instead of:

- lsp.lua
    - omnifunc
- lsp/util.lua
    - extract_completion_items
    - text_document_completion_list_to_complete_items

We could have:

- lsp/completion.lua
    - omnifunc
    - _private_ /local extract_completion_items and text_document_completion_list_to_complete_items
    - future: turn omnifunc sync, add show() for async completion
    - future: Add a way to setup CompleteDone event to handle additionalTextEdits + snipets

**Comment by glepnir** [comment]
plan to add `vim.lsp.autocompletion` and async document preview this pr need wait #24723 . after #24273 done i will back to this pr.

**Comment by max397574** [comment]
> private /local extract_completion_items and text_document_completion_list_to_complete_items

wouldn't it make sense to expose those functions somehow so people can use them for e.g. completion plugins?

**Comment by glepnir** [comment]
item needs to pass the filter and fuzzy match check before it will be added to complete. The logic here doesn't apply to plugins.

**Comment by mfussenegger** [comment]
> I am plan to add vim.lsp.autocompletion

Please don't use `autocompletion` but `completion`. Automatic trigger should (if at all) just be one aspect of the functionality, not the main focus.

**Comment by glepnir** [comment]
I mean vim.lsp.autocompletion can be used in on attach if someone need just a simple completion for lsp and word. there will be no changes to the omnifunc it will still exist. does not break the current behavior.  Just to add an ancillary. Because it's easy to implement and doesn't require too much code .  This is just a thought on my part.

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -2223,6 +2112,18 @@ function M.make_formatting_params(options)
...
+---@param bufnr (integer|nil) Buffer number to resolve. Defaults to current buffer
+---@return integer bufnr
+function M.resolve_bufnr(bufnr)
+  vim.validate({ bufnr = { bufnr, 'n', true } })
+  if bufnr == nil or bufnr == 0 then
+    return api.nvim_get_current_buf()
+  end
+  return bufnr
+end
+
```
Comment (clason) [inline runtime/lua/vim/lsp/util.lua:2126]: We want to get rid of `vim.lsp.util` completely, not dump more stuff in there...

Comment (glepnir) [reply runtime/lua/vim/lsp/util.lua:2126]: okay this is not relate actually ..

Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: Pro tip: the less changes you make in a PR, the faster it is to review it.

Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: This includes _not_ adding any autocompletion (yet). Please keep _this_ PR to a pure refactor, and then make a separate `feat` PR for the autocompletion afterwards (which I'm not 100% certain we want, personally, but it's @mfussenegger's call).

**Comment by justinmk** [comment]
> Because it's easy to implement and doesn't require too much code

That's not a sufficient reason to add LSP-specific autocompletion.

Auto-completion is something we should eventually add to Nvim ([mini-complete](https://github.com/echasnovski/mini.completion) has been working very well for me...), but not in a LSP-specific way.

File: runtime/lua/vim/lsp.lua [hunk 4]
```diff
@@ -26,6 +26,8 @@ local lsp = {
   codelens = require('vim.lsp.codelens'),
   semantic_tokens = semantic_tokens,
   util = util,
+  omnifunc = require('vim.lsp.completion').omnifunc,
```
Comment (justinmk) [inline runtime/lua/vim/lsp.lua:29]: The interface is `vim.lsp.omnifunc`, so this new "completion" module should stay private.
```suggestion
  omnifunc = require('vim.lsp._completion').omnifunc,
```

**Comment by glepnir** [comment]
Because it's small enough. Good enough to use. Most people need it. This is the reason if you want something simple. This is enough. Why install an extra plugin? Better give it a try.  If it is not accepted I will give up this PR because I think it is something worthwhile.

**Comment by glepnir** [comment]
> Generalized autocompletion is on the roadmap. If we add special-case autocompletion, that's extra surface-area that must later be deprecated.

Obviously I thought about these issues. That's why I controlled it to < 200 lines (- code of ominfunc)  and that's why I said it's small enough. And there's nothing to add here. other things can do in plugin or locally.

**Comment by justinmk** [comment]
> That's why I controlled it to < 200 lines (- code of ominfunc)

Code size is only part of the cost calculation. Over time, having many special-case features is a maintenance burden. Since 2014 we have always strived to find ways to generalize mechanisms instead of adding special-case features. It doesn't matter that this particular case in isolation looks small.

**Comment by glepnir** [comment]
I understand what you mean. So I say there is nothing to add here. it only provides basic lsp completion. If anyone wants more it's entirely possible to do it locally and then work together. This is separate.

File: runtime/lua/vim/lsp/util.lua (hunk 3 above)
Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: Is it really too much to ask for to make your enthusiastic changes in smaller, targeted, correctly labelled, PRs?

**Comment by justinmk** [comment]
Something interesting about this PR is `completionProvider.triggerCharacters`. That should be called out in the Problem/Solution description (which says WHY a change is needed). Reviewers should not need to do detective work to figure out why a PR is needed.

Perhaps this suggests that a new option such as 'completetrigger' is needed? That is the kind of [generalized mechanism](https://github.com/neovim/neovim/issues/917#issuecomment-48127449) we are looking for.  cc @echasnovski 

What is needed from Nvim core so that [:help compl-autocomplete](https://neovim.io/doc/user/insert.html#compl-autocomplete) could "just work" with `vim.lsp.omnifunc()`?

**Comment by justinmk** [comment]
@echasnovski thank you! Created https://github.com/neovim/neovim/issues/25670 to track some ideas.

> I am not sure about adding adding a 'completetrigger' option as in "pattern for character after which to trigger completion".

Understandable. Just trying to discover what the key missing pieces needed

> feels too not future proof in case of some complicated async autocompletion would emerge.

What if it were a function that "resolved" the trigger chars on-demand?

**Review by mfussenegger — CHANGES_REQUESTED** [review]
As others have already stated: Please don't mix a fix (https://github.com/neovim/neovim/issues/25177), a refactor and a feature all in the same PR.
It makes it very difficult to review.

Regarding the auto-completion:

I think there are multiple aspects and questions that we'd first need to resolve:

- Do we want this in core? We probably don't want to offer the same scope as something like nvim-cmp does. If we reduce the scope, will people still find it good enough to use it, or would the majority still end up using nvim-cmp? If a reduced scope is okay, how does that look like? E.g. I could upstream https://github.com/mfussenegger/nvim-lsp-compl or parts of it. It takes care of the main LSP aspects.

- If we do add auto-completion, can/should it also work for other completion capabilities in neovim? Omnifunc is one aspect, the LSP specific omnifunc another. What about file completion, etc?

One aspect that's special about LSP is that it also contains a "re-trigger" mechanism. A language server can indicate that the completion result set is incomplete. In that case the client should trigger again after the user entered more characters.

- API design implications: What about the remaining LSP capabilities we're missing? Snippets, `additionalTextEdits`, commands. These are all parts of a completionItem and should be applied when the user selects a candidate. We need to consider the full picture when we design the API. If not, we end up with `auto_completion`, `setup_snippet_expansion`, `setup_commands_handling` and more instead of a single `attach(client, bufnr, opts)`.


---


> What is needed from Nvim core so that :help compl-autocomplete could "just work" with vim.lsp.omnifunc()?

For a plain version, nothing really.
Using trigger characters to trigger the omni-function is possible, in nvim-dap I do something like that:

https://github.com/mfussenegger/nvim-dap/blob/92dc531eea2c9a3ef504a5c8ac0decd1fa59a6a3/lua/dap/ext/autocompl.lua#L15-L46

But to implement the full specification things become more complex because of the re-trigger mechanism. The specification has even context fields, to let the server distinguish between a auto-retrigger and a manual invocation from the user. This breaks a simple "trigger omnifunc" 

And then there's stuff like debounce to avoid server overloading.

There are some things that could be done in core that would make an implementation easier. E.g. https://github.com/neovim/neovim/pull/20671#issuecomment-1288080374

**Comment by echasnovski** [comment]
> What if it were a function that "resolved" the trigger chars on-demand?

I am not sure I understand what "resolved" here would mean.

Broadly speaking, for a working decently generalizable and user-friendly auto-completion functionality in core, there should be:
- Generalizable and user-friendly way to automatically start the completion. This can be as straightforward as a callback that is called on every user input. Inside it should take care of deciding whether to start completion (now or later) and actually starting it.
- Generalizable and user-friendly way to display completion items. There is a `complete()`, but it has downsides described earlier. So a built-in 'completefunc' approach is the only option right with the balance of being usable without too much complications. Also, custom matching would be a welcome addition. I vaguely remember some work in vim/vim for this, but not sure the status of it.

**Comment by glepnir** [comment]
If you really understand vim's completion mechanism. The previous implementation wouldn't have been so bad.

**Comment by glepnir** [comment]
if anyone want fix https://github.com/neovim/neovim/issues/25177  https://github.com/neovim/neovim/pull/17051, https://github.com/neovim/neovim/issues/22803  check how i get start col current is wrong before also wrong . `filtertext` not used in there.

**Pull request closed without merging**